### PR TITLE
Interpreter chooser

### LINKED ISSUE #66: Support multiple paths to different named binaries
Right now, [`autocomplete-python.pythonPath`](https://github.com/sadovnychyi/autocomplete-python/blob/740238ee43222dd25a6ef7d3de5aa7275a7636f4/lib/provider.coffee#L17) and [`autocomplete-python.pythonExecutable`](https://github.com/sadovnychyi/autocomplete-python/blob/740238ee43222dd25a6ef7d3de5aa7275a7636f4/lib/provider.coffee#L18) are separate. The former can have multiple values (searched in order, separated by `;`s). The latter is fixed with a single value.

This makes sense, because `autocomplete-python.pythonPath`'s values are merely prepended to one's `PATH`.

But, I might want to have a setting where I want to degrade from the following:

* `$PROJECT/.venv/bin/python` - uses python2.7
* `python3.4` - default to system-wide Python 3.4 (which is not installed as `.../bin/python`

I can't do that with the current settings.

(See also #25.)

### LINKED ISSUE #112: Autocomplete stops after any minutes on Linux
What information I must give?

### LINKED ISSUE #123: Multiple python processes running with 50% cpu usage
After some light usage of Atom 1.2.4 with autocomplete-python 0.10.6 and py3.4 on win10 x64, multiple python processes appear over time, some of them are running with constant ~50% cpu usage, even if the editor is stopped being used and it is minimized. CPU hogging has been present before with this plugin, but now it also replicates. :stuck_out_tongue:

@sadovnychyi i have seen your recent efforts related to this, first restarting the process after 1k calls, then restarting every 30 minutes, but those measures do nothing when the untraced bug happens to hit just after a few minutes.

Have you thought about asking the help of jedi and atom core devs? Maybe they could offer some advice on how to find and fix the real cause of the bug.

### PR DESCRIPTION
**Pull request by sadovnychyi**
Add new config with optional semicolon separated list of paths to python executables (including executable names), where the first one will take higher priority over the last one. By default autocomplete-python will automatically look for venvs inside of your project and try to use them as well as try to find global python executable.
Use $PROJECT variable for project specific paths to point on executables in virtual environments.

For example: `$PROJECT/env/bin/python3;/usr/bin/python` will fall back on `/usr/bin/python` for projects without venv, but will use env/bin/python3 when possible.

Closes #66

- [x] Try to always use project interpreter if found, without any questions
- [x] Fall back to global one if no local interpreter found
- [x] Cleanup requests queue if there are more than 10 requests in pending state. Closes #112
- [x] Make sure its works for windows. Append .exe prefix for executable search
- [x] Properly kill process on windows. Closes #123

### CODE REVIEW

**Comment by posita** [comment]
Consider eliminating `Python Executable` and `Python Path` in favor of the new `Python executable paths`:

screen shot 2015-12-02 at 19 44 33

**Comment by posita** [comment]
From this https://github.com/sadovnychyi/autocomplete-python/issues/66#issuecomment-161203596:

> Use $PROJECT variable for project specific paths to point on executables in virtual environments.

Based on my understanding of the code, this isn't (strictly speaking) a *variable*, but rather a straight string replacement<sup>1</sup>, correct? In other words `"$PROJECT_FOO"` will yield `".../path/to/project_FOO"`, not `""`, correct?

This isn't an issue if `$PROJECT` is always followed by a path delimiter (e.g., `$PROJECT/`). But you might want to make it clear what is happening in the docs (lest people think you're actually doing variable expansion in that string and think something like `${PROJECT}/bin/python;${HOME}/bin/python` will work).

--

<sup>1</sup> Compare to, e.g., [`expand-var`](https://www.npmjs.com/package/expand-var), which [matches shell-like variable patterns](https://github.com/iexela/expand-var/blob/master/index.js#L9).

**Comment by posita** [comment]
Also, just curious, but why mix `$`-expansion with *semi*colon delimiters? `$` variables seem like UNIX-y (Bourne shell-y?)<sup>1</sup>, while semicolons seem Windoze-y<sup>2</sup>? Maybe that's just me? I guess you could always use `path.delimiter`, but that would require making the documentation platform-specific.

--

<sup>1</sup> As opposed to `%...%` variables in Windows.

<sup>2</sup> As opposed to colons commonly used when evaluating, e.g., the `PATH` environment variable in a Bourne shell.

**Comment by posita** [comment]
As an aside (if you care), I just noticed when evaluating this PR that there is an inconsistent use of Title Case for some of the settings headers (e.g., “***C***ase ***I***nsensitive ***C***ompletion”) and Sentence case for others (e.g., “***U***se *f*uzzy *m*atcher for *c*ompletions”).

File: lib/main.coffee [hunk 1]
```diff
@@ -3,57 +3,77 @@ hyperclickProvider = require './hyperclickProvider'
...
+      order: 3
+      title: 'Python executable paths'
+      description: '''Optional semicolon separated list of paths to python
+      executables (including executable names), where the first one will take
+      higher priority over the last one. By default autocomplete-python will
+      automatically look for venvs inside of your project and try to use them
+      as well as try to find global python executable.
+      Use $PROJECT variable for project specific paths to point on executables
+      in virtual environments.
+      For example: `$PROJECT/env/bin/python3;/usr/bin/python`.
```
Comment (posita) [inline lib/main.coffee]: I think you mean "`$PROJECT/venv/bin/python3;/usr/bin/python`"? (Note the "`venv`" vs "`env`" to be consistent with your next sentence.)

File: lib/main.coffee [hunk 2]
```diff
@@ -3,57 +3,77 @@ hyperclickProvider = require './hyperclickProvider'
...
+      Such config will fall back on `/usr/bin/python` for projects without venv.
```
Comment (posita) [inline lib/main.coffee]: Consider enclose "venv" in backticks for consistency with the previous sentence:

> ... for projects without `venv`.

(See my previous comment.)

File: lib/main.coffee [hunk 3]
```diff
@@ -3,57 +3,77 @@ hyperclickProvider = require './hyperclickProvider'
...
-      title: 'Python executable name'
-      description: 'Set it if default values are not working for you or you want to use specific python version.'
+      order: 3
+      title: 'Python executable paths'
+      description: '''Optional semicolon separated list of paths to python
+      executables (including executable names), where the first one will take
+      higher priority over the last one. By default autocomplete-python will
+      automatically look for venvs inside of your project and try to use them
+      as well as try to find global python executable.
+      Use $PROJECT variable for project specific paths to point on executables
```
Comment (posita) [inline lib/main.coffee]: Consider:

> Use $PROJECT ~~variable~~ *substitution* for project-specific paths ...

See this https://github.com/sadovnychyi/autocomplete-python/pull/125#issuecomment-161509608 for details.

File: lib/main.coffee [hunk 4]
```diff
@@ -3,57 +3,77 @@ hyperclickProvider = require './hyperclickProvider'
...
+      python2 completions for some built-ins.'''
     extraPaths:
       type: 'string'
       default: ''
+      order: 4
       title: 'Extra PATH'
-      description: '''Semicolon separated list of modules to additionally include for autocomplete.
-      You can use $PROJECT variable here to include project specific folders like virtual environment.
+      description: '''Semicolon separated list of modules to additionally
+      include for autocomplete. You can use $PROJECT variable here to include
```
Comment (posita) [inline lib/main.coffee]: Consider:

> ... You can use *the* $PROJECT ~~variable~~ *substitution* here to ...

See this https://github.com/sadovnychyi/autocomplete-python/pull/125#issuecomment-161509608 for details.

**Comment by posita** [comment]
By the, way I think this is great. The above are meant as friendly suggestions only. Please feel free to dismiss or ignore without fear of judgment or reprisal! :grin:

**Comment by sadovnychyi** [comment]
> Also, just curious, but why mix $-expansion with semicolon delimiters? $ variables seem like UNIX-y (Bourne shell-y?)1, while semicolons seem Windoze-y2? Maybe that's just me? I guess you could always use path.delimiter, but that would require making the documentation platform-specific.

Agree with you here, but it seems like more pain if I'll change it for extraPaths as well. Just going to leave it as is.

**Pull request merged**